flowauth/backend/flowauth/token_management.py

```diff
@@ -153,12 +154,19 @@ def add_token(server):
     latest_lifetime = current_user.latest_token_expiry(server)
     if expiry > latest_lifetime:
         raise InvalidUsage("Token lifetime too long", payload={"bad_field": "expiry"})
-    allowed_claims = current_user.allowed_claims(server)
+    # Gotta find all roles that _could_ allow this actio
+    claims = json["claims"]
+    allowed_roles = {role.name: role.is_allowed(claims) for role in current_user.roles}
 
-    current_app.logger.debug("New token request", allowed_claims=allowed_claims)
-    for claim in json["claims"]:
-        if claim not in allowed_claims:
-            raise Unauthorized(f"You do not have access to {claim} on {server.name}")
+    if not any(allowed_roles.values()):
+        raise Unauthorized(
+            f"No roles for {current_user.username} permit the requested scopes."
+        )
+
+    allowed_roles = dict(filter(lambda x: x[1] is True, allowed_roles.items())).keys()
+    current_app.logger.debug(
```

Comment:
I think for the tokens, good to have some debug logging, but should also log to a non-optional audit log that the token was generated?

Reply:
:+1: Already in my local, long with what roles could have been responsible for the token

Reply:
Ah, you mean to a specific `audit_logger` or similar; yes, I agree, I think that's a good idea.